### DISCUSSION TITLE
Any thoughts past "save less versions" about dealing with old/irrelevant version data

### DISCUSSION THREAD

**Discussion by rhanton** [javers/javers#1045] [Q&A]
@bartoszwalacik my team thinks Javers is very useful, but one of our biggest concerns is that the data in our SQL `jv_` tables will get extremely large and unwieldy. Usually there is a point at which old versioning is not needed for applications. This could be time-based (i.e. keep no versions > 1 yr old) or # of versions based, which may make more sense.

I'm curious if you have ever entertained the idea of having a simple way we could run a process and remove "old" snapshots based on some type of query parameters, or if that would cause logical problems when running future JQL queries?

FYI, I did see your answer at https://stackoverflow.com/a/58014407/1571676, but in our case if we're using Spring-Boot and not doing custom control of when snapshots are created, it seems we can get ourselves in trouble.

**Comment by bartoszwalacik** [comment]
Hi, I think that some kind of retention policy for snapshots could be implemented. For example - keep n versions of an object at max, or remove snapshots older than n months, or maybe a  mix of these two.
Generally, removing a consistent sequence of oldest snapshots of an object should be safe, meaning it should not break further JQL queries.
Think we should start from implementing these garbage collector explicitly as a new kind of JQL, leaving a scheduling work for users. Then, in the next version, javers would take responsibility for scheduling. 
We can do it only if you guys are eager to contribute.

**Reply by rhanton** [reply]
I'll see if I can get time to attack the problem from upper management. The developer on my team who has done most of the implementation of javers into one of our services is very bright, so I'd love to have him take a crack at this once he's back from paternity leave. Will create an issue ticket with our attempt at a callable garbage collector once we can take some time to dig into the code base Bartosz.

**Comment by mthrysoe** [comment]
@bartoszwalacik Is this still an option to include in JaVers? 

I'm currently working for a major Danish retailer, who have agreed to set aside a week to make an initial prototype of this feature as we're using JaVers in our new backend platform.

**Reply by bartoszwalacik** [reply]
yes, this issue is open, let me know if you want to contribute, if so, since it's a major change, we should start from a design doc

**Reply by mthrysoe** [reply]
@bartoszwalacik - We would like to contribute this feature.

I was thinking something along the lines of your initial thoughts from 26th of December 2020 as this is exactly what we need.

My plan was to spend the day doing a bit of initial prototyping (throw away), just to get familiar with the code base. Then do a feature description. 

From reading the contributing.md in the project, it's my understanding that the next step is to create an Issue with the proposed design changes. Is that correct?

**Reply by bartoszwalacik** [reply]
Yes, that's correct. It's important to discuss the design first.

**Reply by mthrysoe** [reply]
Your suggestion of making a limited JQL implementation in the initial version sound like a good idea. It limits the scope and proves if the idea is valid and useful. 
 
The plan would be to use this initial limited feature to build the full feature with scheduling in second round.

**Comment by andrea-patricelli** [comment]
Hi all!
@bartoszwalacik is this feature still in progress? Is there any chance to have it in JaVers in future relases? Otherwise do you have any "external" procedure/best practice to apply to a production environment?

**Reply by bartoszwalacik** [reply]
hi @andrea-patricelli , the data retention policy might be added in Javers 8.0 as a premium (paid) feature, stay tuned

**Reply by andrea-patricelli** [reply]
Got it! Thank you  @bartoszwalacik!

**Comment by bartoszwalacik** [comment]
@andrea-patricelli , see https://javers.org/waiting-list
